File: services/signaling/internal/message_storage.py
from lunaricorn.utils.db_manager import *
from .data_types import *
from .signaling_database_manager import *
from datetime import datetime
import json
import time as time_module
from lunaricorn.utils.db_manager import *

class StorageError(Exception):
    pass

class BrokenStorageError(Exception):
    def __init__(self, message):
        self.message = message
        super().__init__(self.message)
    
    def __str__(self):
        return f"BrokenStorageError: {self.message}"


class MessageStorage:
    def __init__(self, db_cfg: DbConfig):
        self.db_cfg = db_cfg
        self.ready = False
        self.logger = logging.getLogger()
        if not db_cfg.valid():
            raise ValueError("invalid db config")
        try:
            db_manager = SignalingDatabaseManager()
            db_manager.initialize(
                host=db_cfg.db_host,
                port=db_cfg.db_port,
                user=db_cfg.db_user,
                password=db_cfg.db_password,
                dbname=db_cfg.db_dbname,
                minconn=1,
                maxconn=3
            )
            db_manager.install_db()
            self.db_manager = db_manager
            self.db_enabled = True
            self.logger.info("Database connection initialized for signaling service")

        except Exception as e:
            self.logger.error(f"Failed to initialize database connection: {e}")
            self.db_enabled = False
            raise BrokenStorageError(f"cannot init storage. reason: {e}")
            


    def create_event(self, event_data: EventData) -> int:
        ctime = datetime.fromtimestamp(event_data.timestamp)
        owner = event_data.source if event_data.source else "ownerless"

        query = """
            INSERT INTO public.signaling_events 
            (type, payload, affected, ctime, owner, tags)
            VALUES (%s, %s, %s, %s, %s, %s)
            RETURNING eid;
        """
        payload_json = json.dumps(event_data.payload) if event_data.payload else None
        affected_json = json.dumps(event_data.affected) if event_data.affected else None
        params = (
            event_data.event_type,
            payload_json,
            affected_json,
            ctime,
            owner,
            event_data.tags
        )
        
        result = self.db_manager.execute_query(
            query=query,
            params=params,
            fetch_one=True
        )
        
        return result[0]

    def get_unique_values(self, field_name):
        if field_name not in ['type', 'affected', 'owner', 'tags']:
            raise ValueError(f"list operation for field {field_name} not supported")

        if field_name in ['type', 'owner']:
            query = f"""
                SELECT DISTINCT {field_name} as f
                FROM public.signaling_events
                WHERE {field_name} IS NOT NULL
                ORDER BY f ASC
            """
            result = self.db_manager.execute_query(query=query, params=None, fetch_one = False,fetch_all = True)
            return [row[0] for row in result]


        elif field_name == 'tags':
            query = """
                SELECT DISTINCT UNNEST(tags) as tag
                FROM public.signaling_events
                WHERE tags IS NOT NULL AND array_length(tags, 1) > 0
                ORDER BY tag ASC
            """
            result = self.db_manager.execute_query(query=query, params=None, fetch_one = False,fetch_all = True)
            return [row[0] for row in result]

        elif field_name == 'affected':
            query = """
                SELECT DISTINCT jsonb_array_elements_text(affected) AS affected_value
                FROM public.signaling_events
                WHERE affected IS NOT NULL 
                AND affected != 'null'::jsonb
                AND jsonb_typeof(affected) = 'array'
                ORDER BY affected_value ASC
            """
            result = self.db_manager.execute_query(query=query, params=None, fetch_one = False,fetch_all = True)
            return [row[0] for row in result]
        
    def find_events(self, timestamp:float, types:list = [], sources:list = [], affected:list = [], tags:list = [], limit:int = 0) -> list:
        conditions = ["ctime >= %s"]
        params = [datetime.fromtimestamp(timestamp)]

        if types:
            placeholders = ",".join(["%s"] * len(types))
            conditions.append(f"type IN ({placeholders})")
            params.extend(types)

        if sources:
            placeholders = ",".join(["%s"] * len(sources))
            conditions.append(f"owner IN ({placeholders})")
            params.extend(sources)

        if affected:
            placeholders = ",".join(["%s"] * len(affected))
            conditions.append(f"affected @> ANY(ARRAY[{placeholders}]::jsonb[])")
            params.extend([f'["{a}"]' for a in affected])

        if tags:
            placeholders = ",".join(["%s"] * len(tags))
            conditions.append(f"tags && ARRAY[{placeholders}]")
            params.extend(tags)


        query = f"""
            SELECT eid, type, payload, affected, ctime, owner, tags
            FROM public.signaling_events 
            WHERE {' AND '.join(conditions)}
            ORDER BY ctime DESC
        """
        if limit > 0:
            query += f" LIMIT {limit}"

        start_time = time_module.perf_counter()
        result = self.db_manager.execute_query(
            query=query,
            params=params,
            fetch_one = False,
            fetch_all = True
        )
        end_time = time_module.perf_counter()
        execution_time = end_time - start_time
        self.logger.info(f"ROBOTS: find_events query - {execution_time} - count: {len(result)} q: {Dbutils.compress_query(query)}, params: {params}")
        events = self.__result_to_event_data_extended_list(result)
        #for event in events:
        #    self.logger.info(f"EID: {event.eid}, Type: {event.event_type}, Timestamp: {event.timestamp}")
        return events

    def __result_to_event_data_extended_list(self, data_list: list) -> List[EventDataExtended]:
        result = []
        for item in data_list:
            event_dict = {
                'eid': item[0],
                'type': item[1],
                'payload': item[2],
                'affected': item[3],
                'ctime': item[4],
                'owner': item[5],
                'tags': item[6]
            }
            event_extended = EventDataExtended.from_dict(event_dict)
            result.append(event_extended)
        
        return result


    def find_events_by_type(self, event_type: str) -> List[EventDataExtended]:
        """
        Find events by type and return them as EventDataExtended objects
        
        Args:
            event_type: The type of events to search for
            
        Returns:
            List of EventDataExtended objects matching the criteria
        """
        query = """
            SELECT eid, type, payload, affected, ctime, owner, tags
            FROM public.signaling_events 
            WHERE type = %s
            ORDER BY ctime DESC;
        """
        
        params = (event_type,)
        
        try:
            results = self.db_manager.execute_query(
                query=query,
                params=params,
                fetch_all=True
            )
            
            events = []
            for row in results:
                # Convert database row to EventDataExtended object
                eid, db_type, payload_json, affected_json, ctime, owner, tags = row
                
                # Parse JSON fields
                payload = json.loads(payload_json) if payload_json else {}
                affected = json.loads(affected_json) if affected_json else None
                
                # Convert datetime to timestamp
                timestamp = ctime.timestamp() if ctime else 0
                
                # Determine source (convert "ownerless" to None)
                source = owner if owner != "ownerless" else None
                
                event = EventDataExtended(
                    eid=eid,
                    event_type=db_type,
                    payload=payload,
                    timestamp=timestamp,
                    source=source,
                    affected=affected,
                    tags=tags
                )
                events.append(event)
            
            return events
            
        except Exception as e:
            self.logger.error(f"Error searching for events of type '{event_type}': {e}")
            return []

File: services/signaling_cpp/app/message_storage.cpp
#include "message_storage.h"

#include <sstream>
#include <iomanip>

namespace lunaricorn
{

namespace
{

constexpr const char* EVENT_SELECT_COLUMNS = R"(
    eid,
    type,
    payload,
    affected,
    extract(epoch from ctime),
    owner,
    tags::text
)";

} // namespace

MessageStorage::MessageStorage(const DbConfig& cfg)
{
    if(!cfg.valid()) {
        throw std::invalid_argument("invalid db config");
    }

    try {
        std::ostringstream conn;
        conn<< "host=" << cfg.dbHost
            << " port=" << cfg.dbPort
            << " user=" << cfg.dbUser
            << " password=" << cfg.dbPassword
            << " dbname=" << cfg.dbDbname;
        sql.open(soci::postgresql,conn.str());
    } catch(const std::exception& e) {
        throw BrokenStorageError(e.what());
    }
}

std::string MessageStorage::json_to_string(const json::value& v)
{
    return json::serialize(v);
}

std::string MessageStorage::tags_to_pg_array(const std::vector<std::string>& tags)
{
    if(tags.empty())
        return "{}";

    std::ostringstream oss;
    oss << '{';
    for(size_t i = 0; i < tags.size(); ++i) {
        if(i > 0)
            oss << ',';
        oss << '"';
        for(char c : tags[i]) {
            if(c == '"' || c == '\\')
                oss << '\\';
            oss << c;
        }
        oss << '"';
    }
    oss << '}';
    return oss.str();
}

std::vector<std::string> MessageStorage::pg_array_to_tags(const std::string& tags_str)
{
    std::vector<std::string> tags;
    if(tags_str.empty() || tags_str == "{}")
        return tags;

    if(tags_str.front() != '{' || tags_str.back() != '}')
        return tags;

    const std::string content = tags_str.substr(1, tags_str.length() - 2);
    if(content.empty())
        return tags;

    size_t start = 0;
    size_t end = 0;
    bool in_quotes = false;
    char quote_char = '\0';

    while(end <= content.length()) {
        if(end < content.length() && (content[end] == '"' || content[end] == '\'')) {
            if(!in_quotes) {
                in_quotes = true;
                quote_char = content[end];
                start = end + 1;
            } else if(content[end] == quote_char) {
                in_quotes = false;
                tags.push_back(content.substr(start, end - start));
                while(end < content.length() && content[end] != ',')
                    ++end;
                start = end + 1;
            }
        } else if(!in_quotes && end < content.length() && content[end] == ',') {
            tags.push_back(content.substr(start, end - start));
            start = end + 1;
        }
        ++end;
    }

    if(start < content.length())
        tags.push_back(content.substr(start));

    return tags;
}

long long MessageStorage::create_event(const StoredEventData& event)
{
    soci::indicator payload_ind =
        event.payload.is_null() ? soci::i_null : soci::i_ok;
    soci::indicator affected_ind =
        event.affected.is_null() ? soci::i_null : soci::i_ok;

    std::string payload = json_to_string(event.payload);
    std::string affected = json_to_string(event.affected);
    std::string owner = event.source ? *event.source : "ownerless";
    std::string tags = tags_to_pg_array(event.tags);

    long long eid;

    soci::statement st = (sql.prepare <<
        R"(
        INSERT INTO signaling_events
        (
            type,
            payload,
            affected,
            ctime,
            owner,
            tags
        )
        VALUES
        (
            :type,
            :payload::jsonb,
            :affected::jsonb,
            to_timestamp(:ctime),
            :owner,
            :tags::text[]
        )
        RETURNING eid
        )",
        soci::into(eid),
        soci::use(event.event_type),
        soci::use(payload, payload_ind),
        soci::use(affected, affected_ind),
        soci::use(event.timestamp),
        soci::use(owner),
        soci::use(tags)
    );

    st.execute(true);
    return eid;
}

std::vector<std::string> MessageStorage::get_unique_values(const std::string& field)
{
    if(field!="type" && field!="owner" && field!="tags" && field!="affected")
        throw std::invalid_argument("unsupported field");

    std::vector<std::string> result;

    if(field=="tags") {
        soci::rowset<std::string> rows = (
            sql.prepare <<
            R"(
            SELECT DISTINCT unnest(tags)
            FROM signaling_events
            ORDER BY 1
            )"
        );
        for(auto& r: rows)
            result.push_back(r);
    } else if(field=="affected") {
        soci::rowset<std::string> rows = (
            sql.prepare <<
            R"(
            SELECT DISTINCT
            jsonb_array_elements_text(affected)
            FROM signaling_events
            ORDER BY 1
            )"
        );
        for(auto& r: rows)
            result.push_back(r);
    } else {
        soci::rowset<std::string> rows = (
            sql.prepare <<
            ("SELECT DISTINCT " + field +
             " FROM signaling_events "
             "WHERE " + field +
             " IS NOT NULL "
             "ORDER BY 1")
        );
        for(auto& r: rows)
            result.push_back(r);
    }

    return result;
}

std::vector<StoredEventDataExtended> MessageStorage::find_events(
    double timestamp,
    const std::vector<std::string>& types,
    const std::vector<std::string>& sources,
    const std::vector<std::string>& affected,
    const std::vector<std::string>& tags,
    int limit
)
{
    (void)affected;

    std::stringstream q;
    q <<
    "SELECT "
    << EVENT_SELECT_COLUMNS <<
    R"(
    FROM signaling_events
    WHERE ctime >= to_timestamp(:ts)
    )";

    if(!types.empty())
        q<<" AND type IN (:types)";
    if(!sources.empty())
        q<<" AND owner IN (:owners)";
    if(!tags.empty())
        q<<" AND tags && :tags::text[]";

    q<<" ORDER BY ctime DESC ";
    if(limit>0)
        q<<" LIMIT "<<limit;

    const std::string query = q.str();
    const std::string tags_pg = tags_to_pg_array(tags);
    const int filter_mask =
        (types.empty() ? 0 : 1) |
        (sources.empty() ? 0 : 2) |
        (tags.empty() ? 0 : 4);

    std::vector<StoredEventDataExtended> result;

    switch(filter_mask) {
    case 0: {
        soci::rowset<soci::row> rows = (sql.prepare << query, soci::use(timestamp));
        for(auto& row: rows)
            result.push_back(row_to_event(row));
        break;
    }
    case 1: {
        soci::rowset<soci::row> rows = (
            sql.prepare << query, soci::use(timestamp), soci::use(types));
        for(auto& row: rows)
            result.push_back(row_to_event(row));
        break;
    }
    case 2: {
        soci::rowset<soci::row> rows = (
            sql.prepare << query, soci::use(timestamp), soci::use(sources));
        for(auto& row: rows)
            result.push_back(row_to_event(row));
        break;
    }
    case 3: {
        soci::rowset<soci::row> rows = (
            sql.prepare << query, soci::use(timestamp), soci::use(types), soci::use(sources));
        for(auto& row: rows)
            result.push_back(row_to_event(row));
        break;
    }
    case 4: {
        soci::rowset<soci::row> rows = (
            sql.prepare << query, soci::use(timestamp), soci::use(tags_pg));
        for(auto& row: rows)
            result.push_back(row_to_event(row));
        break;
    }
    case 5: {
        soci::rowset<soci::row> rows = (
            sql.prepare << query, soci::use(timestamp), soci::use(types), soci::use(tags_pg));
        for(auto& row: rows)
            result.push_back(row_to_event(row));
        break;
    }
    case 6: {
        soci::rowset<soci::row> rows = (
            sql.prepare << query, soci::use(timestamp), soci::use(sources), soci::use(tags_pg));
        for(auto& row: rows)
            result.push_back(row_to_event(row));
        break;
    }
    case 7: {
        soci::rowset<soci::row> rows = (
            sql.prepare << query,
            soci::use(timestamp),
            soci::use(types),
            soci::use(sources),
            soci::use(tags_pg));
        for(auto& row: rows)
            result.push_back(row_to_event(row));
        break;
    }
    }

    return result;
}

std::vector<StoredEventDataExtended> MessageStorage::find_events_by_type(const std::string& type)
{
    soci::rowset<soci::row> rows = (
        sql.prepare <<
        "SELECT "
        << EVENT_SELECT_COLUMNS <<
        R"(
        FROM signaling_events
        WHERE type=:type
        ORDER BY ctime DESC
        )",
        soci::use(type)
    );

    std::vector<StoredEventDataExtended> result;
    for(auto& row: rows)
        result.push_back(row_to_event(row));

    return result;
}

StoredEventDataExtended MessageStorage::row_to_event(soci::row& row)
{
    StoredEventDataExtended e;

    e.eid = row.get<long long>(0);
    e.event_type = row.get<std::string>(1);

    std::string payload = row.get<std::string>(2);
    std::string affected = row.get<std::string>(3);

    e.payload = json::parse(payload);
    if(!affected.empty())
        e.affected = json::parse(affected);

    e.timestamp = row.get<double>(4);

    std::string owner = row.get<std::string>(5);
    if(owner!="ownerless")
        e.source = owner;

    e.tags = pg_array_to_tags(row.get<std::string>(6));
    return e;
}

} // namespace lunaricorn
